File: sdk-javascript/src/recipient/HeaderValidator.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import {
  BadTimestampError,
  BadUuidError,
  MissingHeaderError,
  RecipientCodeMismatchError,
  TimestampOutOfRangeError,
} from '../exceptions/HfcxError.js';
import {
  PROTOCOL_HEADER_API_CALL_ID,
  PROTOCOL_HEADER_CORRELATION_ID,
  PROTOCOL_HEADER_RECIPIENT_CODE,
  PROTOCOL_HEADER_SENDER_CODE,
  PROTOCOL_HEADER_TIMESTAMP,
} from '../protocol/ProtocolHeaders.js';

const UUID_RE = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

export interface HeaderValidatorOptions {
  /** Wall-clock for the timestamp tolerance check. */
  clock?: () => Date;
  /** Tolerance window in ms; default 5 min. */
  timestampToleranceMs?: number;
}

/**
 * Validates the five HFCX protocol headers on an inbound request.
 * Sister to Java's `HeaderValidator`, Python's `HeaderValidator`, and
 * .NET's `HeaderValidator`.
 */
export class HeaderValidator {
  /** Default tolerance: ±5 minutes. */
  static readonly DEFAULT_TIMESTAMP_TOLERANCE_MS = 5 * 60 * 1000;

  private readonly localParticipantCode: string;
  private readonly clock: () => Date;
  private readonly toleranceMs: number;

  constructor(localParticipantCode: string, options: HeaderValidatorOptions = {}) {
    if (!localParticipantCode) throw new TypeError('localParticipantCode is required');
    this.localParticipantCode = localParticipantCode;
    this.clock = options.clock ?? (() => new Date());
    this.toleranceMs =
      options.timestampToleranceMs ?? HeaderValidator.DEFAULT_TIMESTAMP_TOLERANCE_MS;
  }

  validate(headers: Readonly<Record<string, string>>): void {
    if (!headers) throw new TypeError('headers is required');

    require(headers, PROTOCOL_HEADER_SENDER_CODE);
    const recipient = require(headers, PROTOCOL_HEADER_RECIPIENT_CODE);
    const correlationId = require(headers, PROTOCOL_HEADER_CORRELATION_ID);
    const timestamp = require(headers, PROTOCOL_HEADER_TIMESTAMP);
    const apiCallId = require(headers, PROTOCOL_HEADER_API_CALL_ID);

    if (recipient !== this.localParticipantCode) {
      throw new RecipientCodeMismatchError(
        `${PROTOCOL_HEADER_RECIPIENT_CODE} '${recipient}' does not match local '${this.localParticipantCode}'`,
      );
    }

    if (!UUID_RE.test(correlationId)) {
      throw new BadUuidError(
        `${PROTOCOL_HEADER_CORRELATION_ID} is not a valid UUID: '${correlationId}'`,
      );
    }
    if (!UUID_RE.test(apiCallId)) {
      throw new BadUuidError(`${PROTOCOL_HEADER_API_CALL_ID} is not a valid UUID: '${apiCallId}'`);
    }

    const parsed = Date.parse(timestamp);
    if (Number.isNaN(parsed)) {
      throw new BadTimestampError(
        `${PROTOCOL_HEADER_TIMESTAMP} is not valid ISO-8601: '${timestamp}'`,
      );
    }
    const delta = parsed - this.clock().getTime();
    if (delta < -this.toleranceMs || delta > this.toleranceMs) {
      throw new TimestampOutOfRangeError(
        `${PROTOCOL_HEADER_TIMESTAMP} '${timestamp}' is outside ±${this.toleranceMs}ms of the local clock`,
      );
    }
  }
}

function require(headers: Readonly<Record<string, string>>, name: string): string {
  const value = headers[name];
  if (!value) {
    throw new MissingHeaderError(`required protocol header '${name}' is missing or empty`);
  }
  return value;
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/HeaderValidator.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Collections.Generic;
using System.Globalization;
using HealthFlow.Hfcx.Sdk.Exceptions;
using HealthFlow.Hfcx.Sdk.Protocol;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Validates the five HFCX protocol headers on an inbound request.
/// </summary>
/// <remarks>
/// Sister to Java's <c>HeaderValidator</c> and Python's
/// <c>HeaderValidator</c>. Cross-SDK invariants:
/// <list type="bullet">
///   <item>All five headers must be present and non-empty.</item>
///   <item><c>x-hcx-recipient_code</c> must equal the local participant code.</item>
///   <item><c>x-hcx-correlation_id</c> and <c>x-hcx-api-call-id</c>
///     must be valid UUIDs.</item>
///   <item><c>x-hcx-timestamp</c> must parse as ISO-8601 and fall
///     within the configured tolerance (default ±5 min).</item>
/// </list>
/// </remarks>
public sealed class HeaderValidator
{
    /// <summary>Default tolerance: ±5 minutes.</summary>
    public static readonly TimeSpan DefaultTimestampTolerance = TimeSpan.FromMinutes(5);

    private readonly string _localParticipantCode;
    private readonly Func<DateTimeOffset> _clock;
    private readonly TimeSpan _tolerance;

    /// <summary>Construct over the local participant code.</summary>
    public HeaderValidator(
        string localParticipantCode,
        Func<DateTimeOffset>? clock = null,
        TimeSpan? timestampTolerance = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(localParticipantCode);
        _localParticipantCode = localParticipantCode;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        _tolerance = timestampTolerance ?? DefaultTimestampTolerance;
    }

    /// <summary>Validate the headers; throws on first violation.</summary>
    public void Validate(IReadOnlyDictionary<string, string> headers)
    {
        ArgumentNullException.ThrowIfNull(headers);

        var sender = RequireHeader(headers, ProtocolHeaders.SenderCode);
        _ = sender; // sender presence is enough at this layer
        var recipient = RequireHeader(headers, ProtocolHeaders.RecipientCode);
        var correlationId = RequireHeader(headers, ProtocolHeaders.CorrelationId);
        var timestamp = RequireHeader(headers, ProtocolHeaders.Timestamp);
        var apiCallId = RequireHeader(headers, ProtocolHeaders.ApiCallId);

        if (!string.Equals(recipient, _localParticipantCode, StringComparison.Ordinal))
        {
            throw new RecipientCodeMismatchException(
                $"x-hcx-recipient_code '{recipient}' does not match local '{_localParticipantCode}'");
        }

        if (!Guid.TryParseExact(correlationId, "D", out _))
        {
            throw new BadUuidException(
                $"x-hcx-correlation_id is not a valid UUID: '{correlationId}'");
        }

        if (!Guid.TryParseExact(apiCallId, "D", out _))
        {
            throw new BadUuidException(
                $"x-hcx-api-call-id is not a valid UUID: '{apiCallId}'");
        }

        DateTimeOffset parsed;
        try
        {
            parsed = DateTimeOffset.Parse(
                timestamp,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal);
        }
        catch (FormatException)
        {
            throw new BadTimestampException(
                $"x-hcx-timestamp is not valid ISO-8601: '{timestamp}'");
        }

        var delta = parsed - _clock();
        if (delta < -_tolerance || delta > _tolerance)
        {
            throw new TimestampOutOfRangeException(
                $"x-hcx-timestamp '{timestamp}' is outside ±{_tolerance} of the local clock");
        }
    }

    private static string RequireHeader(IReadOnlyDictionary<string, string> headers, string name)
    {
        if (!headers.TryGetValue(name, out var value) || string.IsNullOrEmpty(value))
        {
            throw new MissingHeaderException($"required protocol header '{name}' is missing or empty");
        }

        return value;
    }
}
